### PR TITLE
Switch DeltaQueue pause/resume to use ref counting instead of toggle

### PR DESCRIPTION
**Pull request by curtisman**
- prevents separate part of the code that calls pause/resume to interfere with each other.  
  - This caused trouble when `OpProcessingController` is using pause/resume to control op flow on the container's delta manager, and the runtime is doing it thru the proxy.
- Add test for use of LoaderHeader.pause flag
- Fold creation of `OpProcessingController` with the `LocalTestObjectProvider` so that compat test will use consistent version of the set of loader related objects.  This avoid unnecessary type error and force us to cast to `any`

### CODE REVIEW

File: packages/loader/container-loader/src/deltaQueue.ts [hunk 1]
```diff
@@ -94,14 +94,14 @@ export class DeltaQueue<T> extends TypedEventEmitter<IDeltaQueueEvents<T>> imple
     }
 
     public resume(): void {
-        this.userPause = false;
+        this.userPause--;
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaQueue.ts:97]: assert that it's not negative

File: packages/loader/container-loader/src/deltaQueue.ts [hunk 2]
```diff
@@ -110,7 +110,7 @@ export class DeltaQueue<T> extends TypedEventEmitter<IDeltaQueueEvents<T>> imple
     }
 
     public systemResume(): void {
-        this.sysPause = false;
+        this.sysPause--;
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaQueue.ts:113]: assert that it's not negative

File: packages/test/end-to-end-tests/src/test/compat.spec.ts [hunk 3]
```diff
@@ -54,17 +53,14 @@ describe("loader/runtime compatibility", () => {
             container.on("closed", (error) => containerError = containerError || error !== undefined);
 
             dataObject = await requestFluidObject<TestDataObject>(container as IFluidRouter, "default");
-
```
Comment (vladsud) [inline packages/test/end-to-end-tests/src/test/compat.spec.ts:57]: nit: opProcessingController changes seems unrelated / can be staged as separate RR

File: packages/test/test-utils/src/opProcessingController.ts [hunk 4]
```diff
@@ -10,6 +10,49 @@ import { ILocalDeltaConnectionServer } from "@fluidframework/server-local-server
...
+    public async togglePauseOutbound() {
+        if (!this.outboundPauseP) {
+            this.outboundPauseP = this.deltaManager.outbound.pause();
+        }
+        return this.outboundPauseP;
+    }
+
+    public toggleResumeInbound() {
+        if (this.inboundPauseP) {
+            this.inboundPauseP = undefined;
```
Comment (vladsud) [inline packages/test/test-utils/src/opProcessingController.ts:43]: This implementation assumes that ref-count changes happen synchronously when calling pause() , i.e. ref-counted increase is reflected right away, and thus it's safe to call resume before promise is satisfied.
I feel like all users of pause() should always wait for it to complete before calling resume(). If so, it would be great to assert that requirement here somehow.

File: packages/test/test-utils/src/opProcessingController.ts [hunk 5]
```diff
@@ -167,50 +192,51 @@ export class OpProcessingController {
...
-            if (!this.deltaManagers.has(deltaManager)) {
+    private mapDeltaManagerToggle(deltaMgrs: DeltaManager[]) {
+        if (deltaMgrs.length === 0) {
+            // If no delta managers are provided, process all delta managers in our collection.
+            return Array.from(this.deltaManagerToggles.values());
+        }
+
+        return deltaMgrs.map((deltaManager) => {
+            const toggle = this.deltaManagerToggles.get(deltaManager);
+            if (!toggle) {
```
Comment (vladsud) [inline packages/test/test-utils/src/opProcessingController.ts:215]: can be an assert for readability

**Review by vladsud — APPROVED** [review]

File: packages/test/end-to-end-tests/src/test/compat.spec.ts (hunk 3 above)
Comment (curtisman) [reply packages/test/end-to-end-tests/src/test/compat.spec.ts:57]: Yeah. The problem was discover thru the opProcessingController changes.  So I want to make sure that the solution works for it.

File: packages/test/test-utils/src/opProcessingController.ts (hunk 4 above)
Comment (curtisman) [reply packages/test/test-utils/src/opProcessingController.ts:43]: I am not sure that is always the case.  e.g. (DeltaScheduler.batchEnd)[https://github.com/microsoft/FluidFramework/blob/7f356074c560f83803710ea3320aa9bd41aa3b63/packages/runtime/container-runtime/src/deltaScheduler.ts#L88]

File: packages/test/test-utils/src/opProcessingController.ts (hunk 5 above)
Comment (curtisman) [reply packages/test/test-utils/src/opProcessingController.ts]: done.

File: packages/loader/container-loader/src/deltaQueue.ts (hunk 1 above)
Comment (curtisman) [reply packages/loader/container-loader/src/deltaQueue.ts:98]: Done.

File: packages/loader/container-loader/src/deltaQueue.ts (hunk 2 above)
Comment (curtisman) [reply packages/loader/container-loader/src/deltaQueue.ts:115]: done.

File: packages/loader/container-loader/src/deltaManagerProxy.ts [hunk 6]
```diff
@@ -52,29 +49,19 @@ export class DeltaQueueProxy<T> extends EventForwarder<IDeltaQueueEvents<T>> imp
     }
 
     public async systemPause(): Promise<void> {
-        this.systemPaused = true;
-        return this.queue.pause();
+        return this.pause();
     }
```
Comment (markfields) [inline packages/loader/container-loader/src/deltaManagerProxy.ts:53]: @curtisman - (not that you changed this behavior, just makes me curious to read this code)

Why don't we plumb through systemPause -> systemPause? This must be an important feature of the `DeltaQueueProxy`, but it's pretty strange at face value. A caller using `IDeltaQueue` won't know whether `systemPause` is going to hit the real `DeltaQueue` as a system pause or a user pause...

Comment (markfields) [reply packages/loader/container-loader/src/deltaManagerProxy.ts:53]: Wait actually maybe your change did do something... It looks like the real DeltaQueue logic for system v. user pause is the same as what you deleted here - just keeping track of two different (in name only) reasons to pause, and only resume if both are un-paused.

If someone says this:
- systemPause()
- pause()
- systemResume()

against an `IDeltaQueue` that is actually a `DeltaQueueProxy`, I think it will resume... no?  But before your change it would have not.

Or... never mind this hits the new ref-count implementation and is ok ^_^  In which case get rid of the distinction in the APIs (i.e. remove `systemPause`), right?
